### DISCUSSION TITLE
Writing parallel results - domain decomposition and node numbering

### DISCUSSION THREAD

**Discussion by flotus1** [CGNS/CGNS#298] [CGNStalk]
Looking at the example for Fortran in the documentation (if I am not mistaken, that could use an update, it still uses include 'cgnslib_f.h'???) I have some doubts how to write meshes and later results from an MPI solver with domain decomposition. For the sake of simplicity -and because that's all I need for now- let's limit the problem to Hexa8 unstructured grids with CellCenter results.

What immediately caught my attention is the node/vertex numbering. The example uses a "global" node numbering index. I.e. a node that belongs to two Hexa8 elements from different domains, but at the same position, only appears once in the node coordinate list. And both elements link to the same node index. That's fine within the scope of the example, but I doubt that I can make it work in a real code. Best I can do is probably determine the number of unique nodes in each domain, and shift the node numbering for each domain to avoid overlaps.

The result would be that some nodes at domain boundaries appear more than once in the mesh. That's probably fine as far as the validity of the CGNS file is concerned. Pretty much the same as omitting the node-merging step when writing a mesh in serial.
But from my experience, this WILL cause performance issues when viewing the results in Paraview. On a midrange workstation, I can load and interact with a mesh with several hundred million cells just fine, as long as I "merge" nodes with identical positions in the mesh. But without merging, PV begins to struggle much sooner, I can already tell the difference with 1 million cells. And if I don't have a global node numbering index, all vertices at domain boundaries will remain unmerged.

So here are my questions:
1) Did I overlook a feasible approach to having a global node numbering for all nodes across all domains?
If not:
2) Can i shift the node indices for each domain to avoid overlaps?
3) Are there better solutions to this problem? Maybe something like having a separate zone for each domain? Could that lead to similar performance problems like I saw when writing many time steps to a single CGNS file?
4) Any idea how to get around the performance problems in Paraview when the mesh contains a large number of unmerged vertices? I know there is a filter in Paraview that does node-merging, but that's just shifting the workload to a much less efficient tool.

**Comment by flotus1** [comment]
No comments here? I was able to write a mesh in parallel with local node numbering, single zone. But as expected, ParaView is really struggling to handle meshes with so many unmerged vertices. And the "clean to grid" filter that merges vertices in PV takes quite some time. I can't afford to have another slow filter in my animation pipeline. Let alone one that doesn't seem to scale with the number of cores.

**Reply by brtnfld** [reply]
Just so I understand, are you using a single coordinate numbering and then creating a new section for each partition, much like the example at

https://github.com/obrienadam/Partition-CGNS-Mesh-Example

Then you will not have duplicate nodes.

**Reply by flotus1** [reply]
Not quite. So far all processes write into a single section, just like in the parallel Fortran examples I found in the src directory. The CGNS calls are as follows:
```
call cgp_open_f(trim(adjustl(filename)), CG_MODE_WRITE, cgns_index_file, ier)
call cg_base_write_f(cgns_index_file,basename, 3, 3, cgns_index_base, ier)
isize(1,1) = nverts_unique_total
isize(1,2) = ncells_total
isize(1,3) = 0
call cg_zone_write_f(cgns_index_file, cgns_index_base, zonename, isize, Unstructured, cgns_index_zone, ier)
call cgp_coord_write_f(cgns_index_file, cgns_index_base, cgns_index_zone, RealSingle, 'CoordinateX', index_coord_x, ier)
call cgp_coord_write_f(cgns_index_file, cgns_index_base, cgns_index_zone, RealSingle, 'CoordinateY', index_coord_y, ier)
call cgp_coord_write_f(cgns_index_file, cgns_index_base, cgns_index_zone, RealSingle, 'CoordinateZ', index_coord_z, ier)
call cgp_coord_write_data_f(cgns_index_file, cgns_index_base, cgns_index_zone, index_coord_x, nverts_unique_sum+1_I8B, nverts_unique_sum+nverts_unique, vert_x, ier)
call cgp_coord_write_data_f(cgns_index_file, cgns_index_base, cgns_index_zone, index_coord_y, nverts_unique_sum+1_I8B, nverts_unique_sum+nverts_unique, vert_y, ier)
call cgp_coord_write_data_f(cgns_index_file, cgns_index_base, cgns_index_zone, index_coord_z, nverts_unique_sum+1_I8B, nverts_unique_sum+nverts_unique, vert_z, ier)
nelem_start = ncells_sum+1_I8B
nelem_end = ncells_sum+ncells
nbdyelem = 0_I8B
call cgp_section_write_f(cgns_index_file, cgns_index_base, cgns_index_zone, 'Elem', HEXA_8, 1_I8B, ncells_total, nbdyelem, index_section, ier)
call cgp_elements_write_data_f(cgns_index_file, cgns_index_base, cgns_index_zone, index_section, nelem_start, nelem_end, vertex_index, ier)
call cgp_close_f(cgns_index_file,ier)
```
Each MPI process does that, and only for the grid elements that belong to the domain of this process. Resulting in a single section.
The example you linked seems to do the same, but with a separate section for each "process". Process in quotes because it's a serial program. That's why it can easily write the vertex coordinates in a list with no duplicates.
What I'm trying to do must run MPI parallel. I can't easily create a list of all vertex coordinates across all MPI ranks with no duplicates. So far, I have duplicates at every domain boundary. As far as ParaView is concerned, I don't think it matters whether I have the cell-to-vertex indices in a single section, or in multiple sections for each process. It's the duplicate vertex positions that are causing me a headache.

I was hoping that there is some CGNS functionality that helps me avoid it. Or that someone can point me towards a clever trick for removing those duplicates, without giving up parallelization. I basically came here looking for a shortcut.

**Comment by MicK7** [comment]
Have you tried writing a separate zone for each domain ? It should work well with a ParaView parallel session.

**Reply by flotus1** [reply]
I tried for a while, but could not figure out how to do it in parallel. Could you please give me some hints? Or explain how this solves the problem of duplicate vertices in the coordinate list?
Edit: kinda got it to work by serializing the zone writes. It has no effect on the issue at hand. On the left of this image you can see what PV calls "feature edges". That's the outline of the duplicate nodes, stemming from domain decomposition (3 domains in this example).
![feature_edges](https://user-images.githubusercontent.com/57221355/132951793-b0836c09-6bdf-44ec-9ca8-251d65c68c57.png)
